src/Cognito.re

```diff
@@ -83,4 +165,50 @@ let signUp =
         });
       };
     });
+};
+
+open SignInHelper;
+external makeSignInResponse: 't => signInResponseDecoder = "%identity";
+external toJsDict: 'a => Js.Dict.t(Js.Json.t) = "%identity";
+let signIn = (~username: string, ~password: string, ()) => {
+  let signInParams = authenticationParameters(~username, ~password)->toJsDict;
+  /* create and object of this shape using our `authenticationParameters` type
+     ```json
+     {
+       AuthParameters : {
+         "USERNAME": "idkjs",
+         "PASSWORD": "somepassword",
+       }
+     }
+     ```
+     At this point we know we have a valid js object which can test with `Js.Json.test(signInParams, Object)`
+     Since we know its valid we will change its type to Js.Dict.t(Js.Json.t) which is expected by out params dict. We then use `Js.Dict.set(params, "AuthParameters", Js.Json.object_(signInParams));`to push it to the `AuthParameters` key expected by the cognito client.
+      */
+  let params = Js.Dict.empty();
+  Js.Dict.set(params, "AuthParameters", Js.Json.object_(signInParams));
+  Client.request(`InitiateAuth, params)
+  ->Future.mapOk(resp => {
+      let isErrorResponse = makeResponse(resp);
+      let err = isErrorResponse->__typeGet;
+      let msg = isErrorResponse->messageGet;
+      Js.log2("Raw Response", resp);
+
+      switch (err) {
```

Comment:
See how I'm doing errors now in my most recent dev branch.

I think there's some logic there than can be shared, because all those damn cognito errors makes for a lot of boilerplate.  I think many are shared, though, and since polymorphic variants are composable the mapping operation for any common errors should compose too!

Reply:
FYI, JSCoverter doesn't work there, because I want the variant to carry a payload.

Reply:
Agree on the repeated error logic. Was thinking through. I am going to get the call we have going first then look to refactor the redundancy if you dont get to it first.

Reply:
👍 IMO duplication is always better than a abstraction you're unsure of.